File: library/src/main/java/org/incendo/interfaces/utilities/TitleState.java
package org.incendo.interfaces.utilities;

import net.kyori.adventure.text.Component;

public final class TitleState {

    private Component current;
    private boolean hasChanged = false;

    public TitleState(final Component initialState) {
        this.current = initialState;
    }

    public Component current() {
        this.hasChanged = false;
        return this.current;
    }

    public void current(final Component value) {
        if (this.current.equals(value)) {
            return;
        }

        this.hasChanged = true;
        this.current = value;
    }

    public boolean hasChanged() {
        return this.hasChanged;
    }

}

File: next/src/main/kotlin/org/incendo/interfaces/next/utilities/TitleState.kt
package org.incendo.interfaces.next.utilities

import net.kyori.adventure.text.Component

internal class TitleState(initialState: Component?) {

    internal var current = initialState
        set(value) {
            // Don't update if nothing changed
            if (field == value) return

            hasChanged = true
            field = value
        }
        get() {
            hasChanged = false
            return field
        }

    internal var hasChanged = false
}
